From: Yehuda Sadeh <yehuda@redhat.com>
Date: Thu, 29 Sep 2016 14:07:14 -0700
Subject: rgw: multipart copy part, chunked read

Don't read the entire range from source object, read it in parts.

Signed-off-by: Yehuda Sadeh <yehuda@redhat.com>
(cherry picked from commit 4049e47a0cfc1eef6efd502590b68ba7234589d3)
(cherry picked from commit 335e858defcfd08d5349e4124c9cd8b21cbeeddd)

Resolves: rhbz#1416514
---
 src/rgw/rgw_op.cc | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/rgw/rgw_op.cc b/src/rgw/rgw_op.cc
index 409a3f5..83792c9 100644
--- a/src/rgw/rgw_op.cc
+++ b/src/rgw/rgw_op.cc
@@ -2467,7 +2467,7 @@ int RGWPutObj::get_data(const off_t fst, const off_t lst, bufferlist& bl)
     return ret;
   }
 
-  bl_aux.copy(0, bl_aux.length(), bl);
+  bl.claim_append(bl_aux);
 
   return ret;
 }
@@ -2575,7 +2575,8 @@ void RGWPutObj::execute()
     if (!copy_source) {
       len = get_data(data_in);
     } else {
-      op_ret = get_data(fst, lst, data_in);
+      uint64_t cur_lst = min(fst + s->cct->_conf->rgw_max_chunk_size - 1, lst);
+      op_ret = get_data(fst, cur_lst, data_in);
       if (op_ret < 0)
         goto done;
       len = data_in.length();
